samples/star/src/main/kotlin/com/slack/circuit/star/repo/PetRepository.kt

```diff
@@ -2,49 +2,176 @@
   }
 
-  override suspend fun getAnimal(id: Long): Animal? {
-    if (!this::animals.isInitialized) fetchAnimals()
-    return animals.find { it.id == id }
+  override suspend fun getAnimal(id: Long): DbAnimal? {
+    return withContext(IO) { starDb.starQueries.getAnimal(id).executeAsOneOrNull() }
   }
 
   @Suppress("SwallowedException")
   private suspend fun fetchAnimals() {
-    animals =
-      try {
-        petFinderApi.animals(limit = 100).animals
-      } catch (e: HttpException) {
-        // Sometimes petfinder's API throws 429s for no reason.
-        emptyList()
+    try {
+      val animals = petFinderApi.animals(limit = 100).animals
+      // Do everything in a single transaction for atomicity
+      starDb.transaction {
+        // Delete any not present
+        starDb.starQueries.deleteAllAnimals()
+
+        // Re-populate the DB
+        for ((index, animal) in animals.withIndex()) {
+          starDb.starQueries.updateAnimal(
+            DbAnimal(
+              id = animal.id,
+              sort = index.toLong(),
+              // Names are sometimes all caps
+              name =
+                animal.name.lowercase().replaceFirstChar {
+                  if (it.isLowerCase()) it.titlecase(Locale.US) else it.toString()
+                },
+              url = animal.url,
+              photoUrls = animal.photos.map { it.full }.toImmutableList(),
+              primaryPhotoUrl = animal.photos.firstOrNull()?.medium,
+              tags =
+                listOfNotNull(
+                    animal.colors.primary,
+                    animal.colors.secondary,
+                    animal.breeds.primary,
+                    animal.breeds.secondary,
+                    animal.gender,
+                    animal.size,
+                    animal.status
+                  )
+                  .toImmutableList(),
+              description = animal.description,
+              primaryBreed = animal.breeds.primary,
+              gender = Gender.valueOf(animal.gender.uppercase(Locale.US)),
+              size = Size.valueOf(animal.size.uppercase(Locale.US)),
+              age = animal.age,
+            )
+          )
+        }
+
+        logUpdate("animals")
       }
+    } catch (e: HttpException) {
+      // Sometimes petfinder's API throws 429s for no reason.
+      // TODO retry?
+    }
   }
 
   override suspend fun getAnimalBio(id: Long): String? {
     val animal = getAnimal(id) ?: return null
-    return petFinderApi.animalBio(animal.url)
+    val opId = "bio:$id"
+    val isStale = isOperationStale(opId)
+    val dbBio by lazy(NONE) { starDb.starQueries.getAnimalBio(id).executeAsOneOrNull() }
+    return if (isStale || dbBio == null) {
+      petFinderApi.animalBio(animal.url).also { bio ->
+        // Single transaction to log the operation update with the put
+        starDb.transactionWithResult {
+          logUpdate(opId)
+          starDb.starQueries.putAnimalBio(AnimalBio(id, bio))
+        }
+      }
```

Comment:
Related to my comment above, I think this should be wrapped in `withContext(Dispatchers.IO)` to avoid forcing the responsibility onto the caller (and breaking the abstraction)

Reply:
Done